src/librustdoc/clean/mod.rs

```diff
@@ -1959,33 +1959,44 @@ fn can_elide_trait_object_lifetime_bound<'tcx>(
 #[derive(Debug)]
 pub(crate) enum ContainerTy<'tcx> {
     Ref(ty::Region<'tcx>),
-    Regular { ty: DefId, args: ty::Binder<'tcx, &'tcx [ty::GenericArg<'tcx>]>, arg: usize },
+    Regular {
+        ty: DefId,
+        args: ty::Binder<'tcx, &'tcx [ty::GenericArg<'tcx>]>,
+        has_self: bool,
+        arg: usize,
+    },
 }
 
 impl<'tcx> ContainerTy<'tcx> {
     fn object_lifetime_default(self, tcx: TyCtxt<'tcx>) -> ObjectLifetimeDefault<'tcx> {
         match self {
             Self::Ref(region) => ObjectLifetimeDefault::Arg(region),
-            Self::Regular { ty: container, args, arg: index } => {
+            Self::Regular { ty: container, args, has_self, arg: index } => {
                 let (DefKind::Struct
                 | DefKind::Union
                 | DefKind::Enum
                 | DefKind::TyAlias { .. }
-                | DefKind::Trait
-                | DefKind::AssocTy
-                | DefKind::Variant) = tcx.def_kind(container)
```

Comment:
I've removed `AssocTy` for now (as well those `Some(assoc_ty.def_id)`s) since rustc doesn't handle them either. I have investigated #115379 a bit and compiler just skips them, too. Still unclear if it's a bug / oversight (GATs are younger than object lifetime defaults, so it's possible) or intentional.

Once this question has been resolved, I can properly implement elision in rustdoc for them, too.

`Variant` was never correct. 

Btw, here is the corresponding rustc code:

https://github.com/rust-lang/rust/blob/784916ce245bab8c0a39524da4482684102baf38/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs#L1476-L1488